### ISSUE TITLE
Contract is not truly type safe

### ISSUE THREAD

**Issue by goodhoko** [http4k/http4k#394]
Hello,
I wanted to leverage the 'type safe' contract [http4k claims to provide](https://www.http4k.org/guide/modules/contracts/) to ensure (at compile time) that a handler for given route/method responds with correct response type. However, from the available documentation I was not able to figure out how to do that or whether that's even possible.

Consider following example
```kotlin
data class City(
        val id: Long,
        val name: String,
)

fun getCity(): ContractRoute {
    val openApiSpec = "city/v1/get" meta {
        // type of the response is defined here to be City
        returning(Status.OK, Body.auto<City>().toLens() to City(id = 123, name = "Prague"))
    } bindContract Method.GET

    return openApiSpec to {
        // this is expected to fail at compile type because it does not conform to the spec above
        Response(Status.OK, "booya!")
    }
}
```

Can the expectation be fulfilled? If so, how?

**Comment by goodhoko** [comment]
Spring for example, although using a completely different model, has a way to do this. 
```kotlin
@GetMapping("city/v1/get")
// response type for OpenAPi documentation is read from the return type
open fun getCity(): City {
    // this won't compile
    return "booya!"
}
```

**Comment by daviddenton** [comment]
Yep - we're aware of this and you are of course correct. There is a fine balance between making the API comprehensible (and avoiding magic), which is obviously a little out of whack here.

... that said - there is a bit of a side project that we've been playing with to come up with a completely typesafe routing module which can be bound to create both server and client bindings. It's early days, but we're quite hopeful it might be something that has legs.. :)

**Comment by nlochschmidt** [comment]
You probably know about tAPIr.

https://github.com/softwaremill/tapir

I wonder how much of that could be replicated in Kotlin.

**Comment by daviddenton** [comment]
yep - that's pretty much the general idea in a nutshell. it would be probably fairly easy as a new routing layer unless there are a bunch of implicts or higher kinded types which would need to be converted.

**Comment by jhult** [comment]
Any news on this issue?

**Comment by daviddenton** [comment]
There is no news on this feature request.

**Comment by tjerkw** [comment]
Just to weigh in on this..

Imho an complete type save endpoint could be written as:

```kotlin
@Route("/some/endpoint/{pathParam1}, method=GET)
suspend fun functionName(pathParam1: T1, getParam1: T2): Either<Error, ResultObject) {
   ... 
}
```

This combines something like Arrow for the either type, and a proper serialization lib, such as Kotlin Serialization.

For most endpoints more is not needed. Brging truly function programming to the http world.

I'm building something on top of Ktor for this... However I do like http4k better in the modular functional approach. 
However, if my endpoints look like these functions it doesnt really matter if its http4k or ktor under the hood. Its vanilla Kotlin.
